### ISSUE TITLE
Add settings to remove trailing newlines

### ISSUE THREAD

**Issue by danielcompton** [editorconfig/editorconfig#269]
It would be handy to be able to trim any extra trailing newlines from the end of a file, so there is always just one final newline. I'm not sure if this would be behaviour that would be added to `insert_final_newline` or whether it would be another setting.

**Comment by xuhdev** [comment]
If you set `insert_final_newline=false`, final newlines should be removed. If it is unset or set to something random such as `none`, then it should not be touched.

**Comment by danielcompton** [comment]
Are you sure? The docs on the website say 

> insert_final_newline: set to true to ensure file ends with a newline when saving and false to ensure it doesn't.

They don't seem to say anything about extra trailing new lines?

**Comment by xuhdev** [comment]
Ah, sorry I didn't notice that you meant extra newlines! I can look into this, but this feature is probably limited by what editors provide us.

**Comment by jedmao** [comment]
Yep. I think this comes down to the name `insert_final_newline`, which is an action of inserting a line, not removing 1 or more lines from the end of a document. It's not supposed to support line removals, which means the only way to achieve that would be to introduce a new rule or rename this existing rule into something more meaningful and deprecate the older name.

**Comment by MikeChou** [comment]
Maybe add another setting called 'remove_duplicate_final_newlines'.

The whitespace package of Atom editor combines two settings as one called 'ensure single trailing newline'. But two settings can provide more functions than one (by combinations of these two settings).

**Comment by xuhdev** [comment]
If we use a new name, how would we deal with its compatibility with `insert_final_newline` then? The best way looks to me is add a new legal value `single` to `insert_final_newline` to ensure only one single final new line is there.

**Comment by jedmao** [comment]
@xuhdev, I like that suggestion the best; though, a new rule value like this will most definitely take a while for all the plugins to implement. Just sayin'.

**Comment by xuhdev** [comment]
Yeah, definitely. But we can push the doc first so people can work on it.

**Comment by hax** [comment]
FYI. In Atom editor, this option is called `whitespace.ensureSingleTrailingNewline` and enabled by default.

**Comment by florianb** [comment]
The whitespace package is not compatible to the atom-editorconfig package.

**Comment by fritx** [comment]
Related in vscode:
- https://github.com/Microsoft/vscode/issues/16599
- https://github.com/Microsoft/vscode/issues/30047

**Comment by hax** [comment]
@xuhdev @jedmao Is there any conclusion of this issue?

**Comment by jedmao** [comment]
@xuhdev do we need a vote for `insert_final_newline = single`? I'm a bit concerned about it not being backwards compatible. If people start using `single` before it's implemented then it will effectively behave as an unset, which is the opposite of what people using `single` expect to see. Thoughts?

**Comment by xuhdev** [comment]
@jedmao Can we do another property `trim_extra_final_lines`? This can exist independently with `insert_final_newline`, because some people just want to trim but not have interest in inserting.

**Comment by jedmao** [comment]
I agree. This would be most backwards compatible; however, I'm not 100% on the name. I'm with @fritx on naming it `trim_final_newlines`. This would be better, as `trim_extra_final_lines` doesn't say anything about "new" lines or the newline character sequences.

**Comment by hax** [comment]
Some thought about `insert_final_newline = single` backwards compatible problem:
We may introduce fallback mechanism:
```
insert_final_newline = true

# do not treat unrecognized value as `unset` but just ignore this line, which make last line still effective
insert_final_newline = single
```
Though such fallback mechanism itself breaks backwards compatibility 😔  


Anyway I like `trim_final_newlines` for this case.

**Comment by danielcompton** [comment]
I think `trim_final_newlines` or `trim_trailing_newlines` would both be good. "Trailing" seems to be a common term for extraneous whitespace, whereas "final newlines" is a bit more ambiguous.

**Comment by sierpinski** [comment]
Is there any update on this?

**Comment by Smithienious** [comment]
I also would love to see `trim_trailing_whitespace`. Too bad the devs don't seem interested.

**Comment by norosa** [comment]
I would also like this feature!

**Comment by jessicasmith666** [comment]
@xuhdev Please close this issue.

**Comment by Vigilans** [comment]
VS Code supports `trimFinalNewlines`, along with `insert_final_newline` and `trim_trailing_whitespace` that is already supported by editorconfig. It's a pity that editorconfig do not support this option so I have to add it manually to vscode's workspace level settings (`.vscode/settings.json`) without the guarantee that other development tools would have consistent behavior.

**Comment by daiyam** [comment]
I just found myself in the need of a rule like `trim_final_newline`. I do think it should be added.
If `insert_final_newline` is also enabled, just don't do anything.

**Comment by xuhdev** [comment]
@Vigilans Your issue seems to be the availability of the feature in VSCode. I believe it's being discussed at https://github.com/editorconfig/editorconfig-vscode/issues/153

**Comment by daiyam** [comment]
@xuhdev No it isn't. There is just 1 mention of `trimFinalNewlines` and that issue is all about `trim_trailing_whitespace`.

**Comment by cxw42** [comment]
The specification now declares adherence to SemVer, so we can deal with backward compatibility using the version number.  Since we're still in major version 0, though, anything goes, so maybe we would need to release spec v1 first :D .

I support `insert_final_newline = single`.

**Comment by glen-84** [comment]
Duplicate of #235.

**Comment by phyzical** [comment]
As a workaround you can run something like the following

```
#!/bin/bash

set -e

# Function to process each file
process_file() {
    local file_path="$1"

    echo "Reviewing $file_path"

    # Remove leading newlines
    sed -i '' '1{/^$/d;}' "$file_path"

    # Remove extra newlines within the file
    sed -i '' '/^$/N;/^\n$/D' "$file_path"
}

export -f process_file

ignore_patterns=(
    "read_only"
    ".png"
    ".jpeg"
    ".jpg"
    ".svg"
    ".lock"
    ".zip"
)

files=$(git ls-files -z | tr '\0' '\n')

for pattern in "${ignore_patterns[@]}"; do
    files=$(echo "$files" | grep -v "$pattern")
done

echo "$files" | while IFS= read -r file; do
    process_file "$file" &
done

wait
```

**Comment by cxw42** [comment]
I just ran across [the wiki page about `insert_final_newline`](https://github.com/editorconfig/editorconfig/wiki/Newline-at-End-of-File-Support).

> this feature is probably limited by what editors provide us.

This was also the case at the time we introduced `insert_final_newline`.

>  In Atom editor, this option is called whitespace.ensureSingleTrailingNewline

I like this.  How about:
* new property `ensure_single_trailing_newline = true|false`.  
  * If `true`, the plugin will (if supported by the editor) make sure that the file ends with exactly one of the value indicated by the `end_of_line` property.  (If `end_of_line` is unset, use the platform's native line ending.)
  * If `false` or unset, `insert_final_newline` is used per current behaviour.

**Comment by PanzerFowst** [comment]
I concur with @cxw42 and everyone else who mentioned this as a feature of VS Code.  It would be really nice to have this feature as part of EditorConfig.

**Comment by cxw42** [comment]
@glen-84 Thanks for pointing out the related issue!

**Comment by shynur** [comment]
8 years later...  I can't believe editorconfig STILL donesn't support this feature.  :(
